pysat/_instrument.py

```diff
@@ -1944,109 +2113,270 @@ def bounds(self):
+                                                               ustops,
+                                                               freq=ufreq)
+            else:
+                # instrument has no files
+                self._iter_list = []
+        else:
+            # user provided some inputs
+            starts = np.asarray([start])
+            stops = np.asarray([stop])
+            # ensure consistency if list-like already
+            if len(starts.shape) > 1:
+                starts = starts[0]
+            if len(stops.shape) > 1:
+                stops = stops[0]
+
+            # check equal number of elements
+            if len(starts) != len(stops):
+                estr = ' '.join(('Both start and stop must have the same',
+                                 'number of elements'))
+                raise ValueError(estr)
+
+            # check everything is the same type
+            base = type(starts[0])
+            for lstart, lstop in zip(starts, stops):
+                etype = type(lstop)
+                check1 = not isinstance(lstart, etype)
+                check2 = not isinstance(lstart, base)
+                if check1 or check2:
+                    # Method allows for inputs like inst.bounds = (start, None)
+                    # and bounds will fill the None with actual start or stop.
+                    # Allow for a Nonetype only if length is one.
+                    if len(starts) == 1 and (start is None):
+                        # we are good on type change, start is None, no error
+                        break
+                    elif len(stops) == 1 and (stop is None):
+                        # we are good on type change, stop is None, no error
+                        break
+                    raise ValueError(' '.join(('Start and stop items must all',
                                                'be of the same type')))
-            if isinstance(start[0], str):
+
+            # set bounds based upon passed data type
+            if isinstance(starts[0], str) or isinstance(stops[0], str):
+                # one of the inputs is a string
                 self._iter_type = 'file'
-                self._iter_list = self.files.get_file_array(start, end)
-            elif isinstance(start[0], dt.datetime):
+                # could be (string, None) or (None, string)
+                # replace None with first/last, as appropriate
+                if starts[0] is None:
+                    starts = [self.files[0]]
+                if stops[0] is None:
+                    stops = [self.files[-1]]
+                # default step size
+                if self._iter_step is None:
+                    self._iter_step = 1
+                # default window size
+                if self._iter_width is None:
+                    self._iter_width = 1
+
+                _temp = []
+                self._iter_list = []
+                for _start, _stop in zip(starts, stops):
```

Comment:
Recommend using variable names like `istart`, `istop`, and `temp` to easily differentiate variables in this routine from hidden functions.